File: problems-adapt/0204_primes-below-n/solution_eratosthenes.py
from typing import List, Optional


class Solution:
    def primesBelowN(self, n: int) -> int:
        # No primes strictly below 2.
        if n < 3:
            return 0
        # Sieve of Eratosthenes: assume everything prime, then cross off each
        # prime's multiples in one stroke.
        is_prime = bytearray([1]) * n
        is_prime[0] = is_prime[1] = 0
        i = 2
        # Any composite below n has a factor <= its square root, so nothing
        # new gets marked beyond this bound.
        while i * i < n:
            if is_prime[i]:
                # Zero the multiples of i in one C-level slice assignment,
                # starting at i*i — smaller multiples were crossed off by
                # their smaller factors.
                is_prime[i * i :: i] = bytearray(len(range(i * i, n, i)))
            i += 1
        # Survivors were never a multiple of anything smaller: they are prime.
        return sum(is_prime)

File: problems-adapt/0075_primes-below-n/solution_eratosthenes.js
/**
 * @param {number} n
 * @return {number}
 */
var primesBelowN = function (n) {
    // No primes strictly below 2.
    if (n < 3) return 0;
    // Sieve of Eratosthenes: whatever is never marked composite was not a
    // multiple of anything smaller, so it is prime.
    const isComposite = new Uint8Array(n);
    let count = 0;
    for (let i = 2; i < n; i++) {
        if (!isComposite[i]) {
            count++;
            // Cross off multiples starting at i*i — smaller multiples were
            // marked by their smaller factors.
            if (i * i < n) {
                for (let j = i * i; j < n; j += i) {
                    isComposite[j] = 1;
                }
            }
        }
    }
    return count;
};
